### PR TITLE
Blockchain panics fixes

### PR DESCRIPTION
**Pull request by wlk**
* WIP fixing blockchain.rs panics (moved few possible panics into `Result`, for now just unwrapping them in `insert_block`)
* fixed failing tests
* formatting with editorconfig

### CODE REVIEW

File: ethcore/src/blockchain/blockchain.rs [hunk 1]
```diff
@@ -409,14 +409,14 @@ impl BlockChain {
...
 
 		self.apply_update(ExtrasUpdate {
-			block_hashes: self.prepare_block_hashes_update(bytes, &info),
-			block_details: self.prepare_block_details_update(bytes, &info),
+			block_hashes: self.prepare_block_hashes_update(bytes, &info).unwrap(),
+			block_details: self.prepare_block_details_update(bytes, &info).unwrap(),
 			block_receipts: self.prepare_block_receipts_update(receipts, &info),
 			transactions_addresses: self.prepare_transaction_addresses_update(bytes, &info),
-			blocks_blooms: self.prepare_block_blooms_update(bytes, &info),
+			blocks_blooms: self.prepare_block_blooms_update(bytes, &info).unwrap(),
```
Comment (debris) [inline ethcore/src/blockchain/blockchain.rs]: you should use `try!` here and send this error upstream.

**Comment by debris** [comment]
besides this one function, lgtm

File: ethcore/src/blockchain/blockchain.rs [hunk 2]
```diff
@@ -587,21 +586,21 @@ impl BlockChain {
 
 	/// This functions returns modified blocks blooms.
 	///
-	/// To accelerate blooms lookups, blomms are stored in multiple 
-	/// layers (BLOOM_LEVELS, currently 3). 
+	/// To accelerate blooms lookups, blooms are stored in multiple
```
Comment (NikVolf) [inline ethcore/src/blockchain/blockchain.rs]: though this time small, in general, this change is irrelevant to this pr and should go in the separate one

File: ethcore/src/blockchain/blockchain.rs [hunk 3]
```diff
@@ -627,8 +634,8 @@ impl BlockChain {
 			}
 		};
 
-		modified_blooms.into_iter()
-			.fold(HashMap::new(), | mut acc, (bloom_index, bloom) | {
+		let result = modified_blooms.into_iter()
+		.fold(HashMap::new(), | mut acc, (bloom_index, bloom) | {
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: run-on lines should have a 1 tab indent.

File: ethcore/src/blockchain/blockchain.rs [hunk 4]
```diff
@@ -612,14 +611,22 @@ impl BlockChain {
...
-				let mut blooms: Vec<H2048> = route.iter()
-					.map(|hash| self.block(hash).unwrap())
-					.map(|bytes| BlockView::new(&bytes).header_view().log_bloom())
+				let blocks: Vec<(H256, Option<Bytes>)> = route.iter()
+					.map(|hash| (hash.clone(), self.block(hash)))
 					.collect();
 
+				let mut blooms: Vec<H2048> = Vec::new();
+
+				for block in blocks.into_iter() {
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: i'm sorry to see the lovely functional code be replaced by imperative stuff. is there anything we can do here to make it better?

**Comment by gavofyork** [comment]
whitespace needs correcting. good to merge otherwise.

**Comment by gavofyork** [comment]
[clabot:check]

**Comment by wlk** [comment]
Thanks for the input, I'll address issues soon

File: ethcore/src/blockchain/blockchain.rs (hunk 4 above)
Comment (wlk) [inline ethcore/src/blockchain/blockchain.rs]: @gavofyork Problem is with this line `.map(|hash| self.block(hash).unwrap())` we would like to somehow propagate error information (including `hash` as part of error message). I spent some time with @debris looking at this and it was one reasonable solution. I don't know enough rust to do this in more functional way, but maybe you could suggest something?

**Comment by wlk** [comment]
issues are addressed (except this one which is still outstanding: https://github.com/ethcore/parity/pull/536#discussion-diff-54426250). I'm not sure how to approach this problem

File: ethcore/src/blockchain/blockchain.rs [hunk 5]
```diff
@@ -375,50 +375,56 @@ impl BlockChain {
...
-		}
+		})
 	}
 
 	/// Inserts the block into backing cache database.
 	/// Expects the block to be valid and already verified.
 	/// If the block is already known, does nothing.
-	pub fn insert_block(&self, bytes: &[u8], receipts: Vec<Receipt>) {
+	/// Returns Option<H256> - newly inserted block hash if insert was successful
+	pub fn insert_block(&self, bytes: &[u8], receipts: Vec<Receipt>) -> Result<Option<H256>, Error> {
```
Comment (arkpar) [inline ethcore/src/blockchain/blockchain.rs]: Is it guaranteed that the blockchain remains in consistent state after returning an error from this function?

File: ethcore/src/blockchain/blockchain.rs (hunk 4 above)
Comment (tomusdrw) [inline ethcore/src/blockchain/blockchain.rs]: That might be usefull:
`One of the keys to collect()'s power is that many things you might not think of as 'collections' actually are. For example, a String is a collection of chars. And a collection of Result<T, E> can be thought of as single [Result<Collection<T>, E>]. See the examples below for more.`
http://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect

Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: more of a wish - i can't see any easy way of doing it either :(

**Comment by arkpar** [comment]
After an internal discussion we decided to take on the following rules for error handling:
There are two kinds of errors.
1. Module invariants (a.k.a Programmer’s mistakes, assertions). These should never happen by module design and if they do happen, they should panic as soon as possible. Also there should be a single place of function to check for such error not to litter the code with `expect(“This should never happen)`.
2. Normal errors caused by invalid input, resource exhaustion, file corruption, other **external** reasons. These should be returned as `Error` and in general can be handled.
There should be no `FatalError` returned from the module interface. In case of blockchain it’s a bit complicated. The situation when a block is in the database, but the details are not can be caused by two reasons: either there is a database corruption, or a programmer's mistake. The first one should be an error and can be handled by rebuilding the database for example. The second one should be a panic. The error handling here should start from the lowest level - the database. The error returned from the database should be treated as an error. Missing data should panic.

**Comment by gavofyork** [comment]
Alternative branch: wlk-blockchain_panics_fixes

**Pull request closed without merging**